### ISSUE TITLE
afterSave for new models flushes unsaved data

### ISSUE THREAD

**Issue by yujin1st** [yiisoft/yii2#15875]
After #13567 AR now flushed relational data, so there is new issue:

### What steps will reproduce the problem?
1. Create simple app with Order and OrderItem models.
2. Load data in controller, create relational OrderItem records and populate relation 'items'
```php
// In Order 
public function load($data, $formName = null) {
  if (!parent::load($data, $formName)) return false;
  $this->loadItems($data);
  return true;
}

/// loading and populating data
private function loadItems($data) {
  if (!isset($data['items'])) {
    return;
  }
  $items = $this->items;
  foreach ($data['items'] as $info) {
    $itemId = $info['itemId'];
    if (isset($items[$itemId])) {
      $items[$itemId]->load($info, '');
    } else {
      $item = new OrderItem([]);
      $item->load($info, '');
      $items[$itemId] = $item;
    }
  }
  $this->populateRelation('items',$items);
}


```

3. Validate, do something in beforeSave method and event
3.1 Validate in Order model
3.2 For example warehouse module can check item quantity before order creating in event
4. Save items in afterSave
```php
public function afterSave($insert, $changedAttributes) {
  $this->saveItems();
  parent::afterSave($insert, $changedAttributes);
}

```

5. Do something with new items in afterSave method and event
 
### What is the expected result?
All works fine, order data saved, relational data saved too

### What do you get instead?
After #13567 items relation with unfinished data is flushed, as Order model changes it's id, so you lose data

### Additional info
It simple workflow allows to validate and process relational form data as ActiveRecord models with all it features and controllers remains thing. And you could use scenarios for complex relations, so models can be easily tested.

I'm afraid this issue will be closed due contradiction with updated framework design, but i have a multiply applications that follows this workflow, so i don't know what to do ;(((

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.14

**Comment by samdark** [comment]
@markmarco16, @rob006, @Kolyunya thoughts?

**Comment by cebe** [comment]
which operation causes the reset of relational data? A one-many relation has no attribute in the order that should cause resetting the relation when modified.

**Comment by samdark** [comment]
Indeed, that should not be related to #13567 as there are no relation attributes in the model you modify... @yujin1st are you sure your problem caused by #13567?

**Comment by yujin1st** [comment]
I'm sure. Debugging leads to this code:

```php
ActiveRecord class:

protected function insertInternal($attributes = null){
//...
//lines 518
        foreach ($primaryKeys as $name => $value) {
            $id = static::getTableSchema()->columns[$name]->phpTypecast($value);
            $this->setAttribute($name, $id);
            $values[$name] = $id;
// ...
        }

BaseActiveRecord class: 
public function setAttribute($name, $value){
//...
//line 521
  $this->resetDependentRelations($name);
//...
}

```

internal property ***_relationsDependencies***  is set when loading items, and this line is executed to preload old saved data:
```
  $items = $this->items; 
```

At this moment relation is not set, so ***__get*** method executes, 
and 
```php
BaseActiveRecord class: 
 public function __get($name){
///.. 296 line
 if ($value instanceof ActiveQueryInterface) {
            $this->setRelationDependencies($name, $value);
            return $this->_related[$name] = $value->findFor($name, $this);
        }
}
```
=======
now after debugging i think i can avoid this issue by adding ***if***:
```php
$items =  $this->isNewRecord ? [] : $this->items;
```

But will be this correct resolution? 
It means every time you access to relation you should keep in mind, if model is new, if it's foreign keys can be updated by this, and after all this means it's a lot places to rewrite.. =(

**Comment by yujin1st** [comment]
And i think i found another interesting case: 
What happens if you load items with lazy loading ($query->with()) , so relation is populated  in ***\yii\db\ActiveRelationTrait::populateRelation***, so ***setRelationDependencies*** method is not called and internal property is not set too.

**Comment by samdark** [comment]
Hmm... let's get back to Order model. It has one to many relation so there's no column in Order that makes a relation. Is that correct? What action with the model triggers resetting a relation?

**Comment by yujin1st** [comment]
Doesn't ***id*** column compose relation? 
$model->save()  reset relation: save()->insert->primary key ***id*** updates (from null to value)->setAttribute()->resetDependentRelations() -> relation resets;

**Comment by samdark** [comment]
Do you change id?

**Comment by samdark** [comment]
So the case is only about not yet saved model, right?

**Comment by yujin1st** [comment]
Yes, when id is changing after saving new model

**Comment by samdark** [comment]
I guess in this case we should not reset relations and your extra condition may fix it.

**Comment by mg-code** [comment]
Same issue here.
I was forced to refactor many forms to fix this bug.

I really like the idea that we can populate relation manually and then use related models same way for create/update actions. After this merge I get empty relations after saving parent model.

**Comment by samdark** [comment]
@md-code do you have other ideas (not a flag to turn it on/off) on how to fix that?

**Comment by mg-code** [comment]
@samdark one solution would be to add a new parameter to function `populateRelation` which indicates whether to keep relation in sync. By default it would be false, but in all other places (e.g. ActiveQuery) we can set it to true.

Imho, probably not the best solution, but better than current.

**Comment by yujin1st** [comment]
May be add some behaviour to trace the changes if needed?

**Comment by yujin1st** [comment]
sorry, i missed that fact, that we can't override methods in behaviours.

**Comment by mg-code** [comment]
It would be better to create a behavior that resets relation, not visa versa.
Because, currently this merge breaks backward compatibility.

We could create an event (e.g. AFTER_ATTRIBUTE_CHANGE) and for those who need a functionality of reseting relations could use some behavior.

**Comment by rob006** [comment]
There is already a PR with fix: #16085. What is wrong with it?

**Comment by mg-code** [comment]
I think it might work.

I thought that ActiveQuery is calling `populateRelation` to fill relations. Now I see that it is not and there is no need for an additional parameter.

**Comment by samdark** [comment]
@mg-code it's in. Tests are fine so I think it will work well. Please test with master.

**Comment by mg-code** [comment]
@samdark yes, it works now. Thanks!

**Comment by cebe** [comment]
https://github.com/yiisoft/yii2/pull/16085#discussion_r183968666

> the tests in https://github.com/yiisoft/yii2/pull/13618/files#diff-5c2b59d39a2ea1dc5443fb98d1325d11 do not cover eager loading. Pretty sure this gets broken by this change.

**Comment by cebe** [comment]
> I thought that ActiveQuery is calling populateRelation to fill relations. Now I see that it is not and there is no need for an additional parameter.

it does in case of eager loading:
 https://github.com/yiisoft/yii2/blob/c3c982068dc21697e1c167bf2d20e55813aa4f0d/framework/db/ActiveRelationTrait.php#L253

**Comment by mg-code** [comment]
@cebe Yes, you are right. So additional parameter can be a solution.

**Comment by cebe** [comment]
or a separate method, e.g. populateRelationInternal()

**Comment by mg-code** [comment]
I was thinking about that, but came to idea that internal methods are protected or private, but we need to call this method from outside.

**Comment by yujin1st** [comment]
is any possibility to use this fix in stable version?
we can't use any version above 2.0.14  because of it =(

**Comment by shirase** [comment]
Issue already fixed in master.

**Comment by samdark** [comment]
Yes. Will tag next release as soon as I'll have a bit more time.
